app/grandchallenge/profiles/views.py

```diff
@@ -197,3 +206,94 @@ def form_invalid(self, form):
         # and display an error message
         messages.add_message(self.request, messages.ERROR, "Incorrect token.")
         return response
+
+
+class SubscriptionUpdate(SuccessMessageMixin, UserPassesTestMixin, UpdateView):
+    model = UserProfile
+    form_class = SubscriptionPreferenceForm
+    template_name = "profiles/subscription_form.html"
+    success_message = "You successfully updated your subscription preferences."
+    raise_exception = True
+
+    def test_func(self):
+        try:
+            token = self.validate_token()
+        except BadSignature:
+            return False
+
+        if (
+            not get_user_model()
+            .objects.filter(username__iexact=token, is_active=True)
+            .exists()
+        ):
+            return False
+
+        return True
+
+    def validate_token(self):
+        return Signer(salt=UNSUBSCRIBE_SALT).unsign(self.kwargs.get("token"))
+
+    def get_object(self):
+        return get_object_or_404(
+            UserProfile, user__username=self.validate_token()
+        )
+
+    def form_valid(self, form):
+        if (
+            self.request.user.is_authenticated
+            and self.object.user != self.request.user
+        ):
+            update_verification_user_set.signature(
+                kwargs={
+                    "usernames": [
+                        self.request.user.username,
+                        self.object.user.username,
+                    ]
+                }
+            ).apply_async()
```

Comment:
1. True, I hadn't though about GET. I think it makes sense to do that always. I'm unsure where to move this check to though. Moving it to `test_func` would be straight forward, but then test_func does more than just a permission checking. Other methods that run before rendering the form are `get_context_data` and `get_form_kwargs` - semantically, neither of those fit the purpose either. So I'm tempted to add it to test_func. What do you think?

Reply:
I think that doing it in `test_func` is probably fine as it's an action we want to take in response to a dodgy permissions check. Otherwise, you would do it in `dispatch` or `get`, but then you might end up doing double work elsewhere.